# Assignment 2: Fake Job Posting Detection

## Dataset: Fake Job Postings

This notebook contains the analysis for detecting fraudulent job postings using machine learning.


In [2]:
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Set display options
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', 100)

print("Libraries imported successfully!")


Libraries imported successfully!


## Step 1: Load and Explore the Data


In [3]:
# Load the dataset
df = pd.read_csv('fake_job_postings.csv')

print(f"Dataset shape: {df.shape}")
print(f"\nColumn names:")
print(df.columns.tolist())
print(f"\nFirst few rows:")
df.head()


Dataset shape: (17880, 18)

Column names:
['job_id', 'title', 'location', 'department', 'salary_range', 'company_profile', 'description', 'requirements', 'benefits', 'telecommuting', 'has_company_logo', 'has_questions', 'employment_type', 'required_experience', 'required_education', 'industry', 'function', 'fraudulent']

First few rows:


job_id                                      title            location  \
0       1                           Marketing Intern    US, NY, New York   
1       2  Customer Service - Cloud Video Production      NZ, , Auckland   
2       3    Commissioning Machinery Assistant (CMA)       US, IA, Wever   
3       4          Account Executive - Washington DC  US, DC, Washington   
4       5                        Bill Review Manager  US, FL, Fort Worth   

  department salary_range  \
0  Marketing          NaN   
1    Success          NaN   
2        NaN          NaN   
3      Sales          NaN   
4        NaN          NaN   

                                                                                       company_profile  \
0  We're Food52, and we've created a groundbreaking and award-winning cooking site. We support, con...   
1  90 Seconds, the worlds Cloud Video Production Service.90 Seconds is the worlds Cloud Video Produ...   
2  Valor Services provides Workforce Solutions that meet the needs of companies across the Private ...   
3  Our passion for improving quality of life through geography is at the heart of everything we do....   
4  SpotSource Solutions LLC is a Global Human Capital Management Consulting firm headquartered in M...   

                                                                                           description  \
0  Food52, a fast-growing, James Beard Award-winning online food community and crowd-sourced and cu...   
1  Organised - Focused - Vibrant - Awesome!Do you have a passion for customer service? Slick typing...   
2  Our client, located in Houston, is actively seeking an experienced Commissioning Machinery Assis...   
3  THE COMPANY: ESRI – Environmental Systems Research InstituteOur passion for improving quality of...   
4  JOB TITLE: Itemization Review ManagerLOCATION: Fort Worth, TX                                   ...   

                                                                                          requirements  \
0  Experience with content management systems a major plus (any blogging counts!)Familiar with the ...   
1  What we expect from you:Your key responsibility will be to communicate with the client, 90 Secon...   
2  Implement pre-commissioning and commissioning procedures for rotary equipment.Execute all activi...   
3  EDUCATION: Bachelor’s or Master’s in GIS, business administration, or a related field, or equiva...   
4  QUALIFICATIONS:RN license in the State of TexasDiploma or Bachelors of Science in Nursing, requi...   

                                                                                              benefits  \
0                                                                                                  NaN   
1  What you will get from usThrough being part of the 90 Seconds team you will gain:experience work...   
2                                                                                                  NaN   
3  Our culture is anything but corporate—we have a collaborative, creative environment; phone direc...   
4                                                                                Full Benefits Offered   

   telecommuting  has_company_logo  has_questions employment_type  \
0              0                 1              0           Other   
1              0                 1              0       Full-time   
2              0                 1              0             NaN   
3              0                 1              0       Full-time   
4              0                 1              1       Full-time   

  required_experience required_education                   industry  \
0          Internship                NaN                        NaN   
1      Not Applicable                NaN  Marketing and Advertising   
2                 NaN                NaN                        NaN   
3    Mid-Senior level  Bachelor's Degree          Computer Software   
4    Mid-Senior level  Bachelor's Degree     Hospital

In [4]:
# Basic information about the dataset
print("Dataset Info:")
print(df.info())
print("\n" + "="*50)
print("\nMissing values:")
print(df.isnull().sum())
print("\n" + "="*50)
print("\nTarget variable distribution:")
print(df['fraudulent'].value_counts())
print(f"\nFraudulent rate: {df['fraudulent'].mean():.2%}")


Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17880 entries, 0 to 17879
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   job_id               17880 non-null  int64 
 1   title                17880 non-null  object
 2   location             17534 non-null  object
 3   department           6333 non-null   object
 4   salary_range         2868 non-null   object
 5   company_profile      14572 non-null  object
 6   description          17879 non-null  object
 7   requirements         15184 non-null  object
 8   benefits             10668 non-null  object
 9   telecommuting        17880 non-null  int64 
 10  has_company_logo     17880 non-null  int64 
 11  has_questions        17880 non-null  int64 
 12  employment_type      14409 non-null  object
 13  required_experience  10830 non-null  object
 14  required_education   9775 non-null   object
 15  industry             12977 non-null  ob

## Step 2: Simple Data Cleaning

Based on the initial exploration, we'll perform basic cleaning operations.


In [5]:
# Create a copy for cleaning
df_clean = df.copy()

# Check for duplicate rows
print(f"Number of duplicate rows: {df_clean.duplicated().sum()}")

# Remove duplicates if any
if df_clean.duplicated().sum() > 0:
    df_clean = df_clean.drop_duplicates()
    print(f"Removed duplicates. New shape: {df_clean.shape}")
else:
    print("No duplicates found.")


Number of duplicate rows: 0
No duplicates found.


In [6]:
# Handle missing values - fill empty strings with NaN for consistency
text_columns = ['title', 'location', 'department', 'company_profile', 
                'description', 'requirements', 'benefits', 'employment_type',
                'required_experience', 'required_education', 'industry', 'function']

for col in text_columns:
    if col in df_clean.columns:
        # Replace empty strings with NaN
        df_clean[col] = df_clean[col].replace('', np.nan)

print("Missing values after cleaning:")
print(df_clean[text_columns].isnull().sum())


Missing values after cleaning:
title                      0
location                 346
department             11547
company_profile         3308
description                1
requirements            2696
benefits                7212
employment_type         3471
required_experience     7050
required_education      8105
industry                4903
function                6455
dtype: int64


In [7]:
# Check data types and ensure target variable is correct
print("Data types:")
print(df_clean.dtypes)
print(f"\nTarget variable (fraudulent) type: {df_clean['fraudulent'].dtype}")
print(f"Target variable unique values: {df_clean['fraudulent'].unique()}")

# Ensure target is integer (should already be, but just to be safe)
df_clean['fraudulent'] = df_clean['fraudulent'].astype(int)

print(f"\nFinal dataset shape: {df_clean.shape}")
print(f"Final fraudulent rate: {df_clean['fraudulent'].mean():.2%}")


Data types:
job_id                  int64
title                  object
location               object
department             object
salary_range           object
company_profile        object
description            object
requirements           object
benefits               object
telecommuting           int64
has_company_logo        int64
has_questions           int64
employment_type        object
required_experience    object
required_education     object
industry               object
function               object
fraudulent              int64
dtype: object

Target variable (fraudulent) type: int64
Target variable unique values: [0 1]

Final dataset shape: (17880, 18)
Final fraudulent rate: 4.84%


---

# Section 1: Identify the Predictive Task

## Task 1: Binary Classification - Detecting Fraudulent Job Postings

Predict whether a job posting is fraudulent (1) or legitimate (0).

- Input: Job posting features (text: description, requirements; categorical: location, industry; binary: has_company_logo, has_questions)
- Output: Binary label (0 = real, 1 = fraudulent)
- Dataset: 17,880 samples, 4.84% fraudulent (imbalanced)

## Task 2: Multiclass Classification - Predicting Employment Type

Predict the employment type of a job posting (Full-time, Part-time, Contract, etc.).

- Input: Same job posting features as Task 1
- Output: Employment type category (multiclass)
- Dataset: 14,409 samples with employment_type labels (after removing missing values)


## How We Will Evaluate Our Models

### Task 1 (Binary Classification - Fraudulent Detection):

Given class imbalance (4.84% fraudulent), we use:
- **Precision:** Minimizes false positives (flagging real jobs as fake)
- **Recall:** Maximizes catching fraud (missing fake jobs)
- **F1-Score:** Balances precision and recall
- **ROC-AUC:** Model discrimination across all thresholds (0.5 = random, 1.0 = perfect)
- **Accuracy:** Reference metric (note: "always predict real" gets 95.16% accuracy but 0% recall)

**Evaluation Protocol:**
- 80/20 stratified train/test split
- 5-fold stratified cross-validation for model selection
- Report all metrics (Precision, Recall, F1, ROC-AUC, Accuracy)

### Task 2 (Multiclass Classification - Employment Type):

- **Accuracy:** Overall correctness
- **Macro F1-Score:** Average F1 across all classes (handles class imbalance)
- **Per-Class Precision/Recall:** Performance for each employment type
- **Confusion Matrix:** Error analysis across classes

**Evaluation Protocol:**
- 80/20 stratified train/test split
- 5-fold stratified cross-validation for model selection
- Report accuracy, macro F1, and per-class metrics


## Relevant Baselines for Comparison

Following course examples (similar to review rating task using Linear Regression baseline before Lasso/Ridge):

### Task 1 (Binary Classification - Fraudulent Detection):

**Baseline 1: Random Classifier**
- Randomly predicts 0 or 1 (50/50)
- Expected: ~50% accuracy, ROC-AUC ~0.50

**Baseline 2: Majority Class Classifier**
- Always predicts real (0)
- Expected: 95.16% accuracy, 0% recall, ROC-AUC 0.50

**Baseline 3: Logistic Regression (No Regularization)** - **MAIN BASELINE**
- Logistic Regression without regularization (C=1.0)
- **Course Relevance:** Main baseline from class (similar to Linear Regression baseline before Lasso/Ridge)
- Features: All features (text via TF-IDF, binary, categorical)
- Purpose: Baseline ML performance before adding regularization

**Comparison:** Our models (Logistic Regression with L1/L2 regularization) should outperform this baseline, demonstrating regularization benefits.

### Task 2 (Multiclass Classification - Employment Type):

**Baseline 1: Random Classifier**
- Randomly predicts employment type
- Expected: ~1/N accuracy (where N = number of classes)

**Baseline 2: Majority Class Classifier**
- Always predicts most common employment type
- Expected: Accuracy = proportion of majority class

**Baseline 3: Logistic Regression (No Regularization)** - **MAIN BASELINE**
- Logistic Regression (multiclass) without regularization (C=1.0)
- **Course Relevance:** Main baseline from class (similar to Linear Regression baseline)
- Features: All features (text via TF-IDF, binary, categorical)
- Purpose: Baseline ML performance before adding regularization

**Comparison:** Our models (Logistic Regression with L1/L2 regularization, Random Forest) should outperform this baseline.


## How We Will Assess Prediction Validity

We use validation strategies to ensure predictions are reliable and generalizable:

1. **Train/Test Split with Stratification**
   - **Task 1:** 80/20 split, stratified by `fraudulent` label (maintains 4.84% rate)
   - **Task 2:** 80/20 split, stratified by `employment_type` (maintains class distribution)
   - Purpose: Evaluate on unseen data, prevent overfitting

2. **Cross-Validation for Model Selection**
   - 5-fold stratified cross-validation on training set
   - Purpose: Select best hyperparameters (C values for regularization)
   - Use: Compare models, regularization strengths, feature combinations

3. **Confusion Matrix Analysis**
   - **Task 1:** Analyze TP, TN, FP, FN (identify if too conservative or aggressive)
   - **Task 2:** Analyze errors across employment type classes

4. **Learning Curves**
   - Plot training vs validation performance
   - Detect overfitting (large gap) or underfitting (both low)

5. **Feature Importance Analysis**
   - Analyze coefficients (logistic regression) or feature importances (random forest)
   - Validate features align with domain knowledge

6. **ROC Curve Analysis** (Task 1 only)
   - Plot TPR vs FPR across all thresholds
   - Compare models and select optimal threshold

7. **Performance Stability**
   - Evaluate on multiple random splits
   - Ensure consistent performance

These strategies ensure our models' predictions are reliable, generalizable, and not overfitted.
